METRON-1480 Add yarn as default build tool for the frontend

xyztdanid4 opened the pull request:
Updated the build tool to yarn for the frontend project to avoid npm instability during dependency adding.

### For all changes:
- [X] Is there a JIRA ticket associated with this PR? If not one needs to be created at [https://issues.apache.org/jira/browse/METRON-1480](https://issues.apache.org/jira/browse/METRON/?selectedTab=com.atlassian.jira.jira-projects-plugin:summary-panel).
- [X] Does your PR title start with METRON-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.
- [ ] Has your PR been rebased against the latest commit within the target branch (typically master)?

### For code changes:
- [ ] Have you included steps to reproduce the behavior or problem that is being changed or addressed?
- [ ] Have you included steps or a guide to how the change may be verified and tested manually?
- [X] Have you ensured that the full suite of tests and checks have been executed in the root metron folder via:
  ```
  mvn -q clean integration-test install && dev-utilities/build-utils/verify_licenses.sh 
  ```

- [ ] Have you written or updated unit tests and or integration tests to verify your changes?
- [X] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [X] Have you verified the basic functionality of the build by building and running locally with Vagrant full-dev environment or the equivalent?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered by building and verifying the site-book? If not then run the following commands and the verify changes via `site-book/target/site/index.html`:

  ```
  cd site-book
  mvn site
  ```

#### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.
It is also recommended that [travis-ci](https://travis-ci.org) is set up for your personal repository such that your branches are built there before submitting a pull request.

simonellistonball commented:
Have the dependencies, License and Notices files been updated?

xyztdanid4 commented:
@simonellistonball I am not sure. Probably no.

cestella commented on metron-interface/metron-alerts/package-lock.json:
```diff
@@ -1,6427 +0,0 @@
-{
```
I suspect the build failure is due to a lack of a license on this file.  Since JSON cannot take comments, you'll need to adjust the top level pom to add an exclusion for `**/package-lock.json` [here](https://github.com/apache/metron/blob/master/pom.xml#L304)  Just make a quick comment above it and note that this is yarn collateral.

xyztdanid4 replied:
I think that was the problem!
Thanks. @cestella

merrimanr commented on metron-interface/metron-config/package.json:
```diff
@@ -33,6 +28,11 @@
     "@angular/platform-browser": "2.0.0",
     "@angular/platform-browser-dynamic": "2.0.0",
     "@angular/router": "3.0.0",
+    "@types/ace": "0.0.32",
```
Were there any real changes made to this file?  It looks like some dependencies were simply rearranged.  I do see the addition of `"main": "index.js"`.  What is the purpose of that?

xyztdanid4 replied:
No there were no changes other than rearrange of the dependencies.
About the main attribute, in our case it will do nothing. It defines the entry point of the application if we would SystemJS or requireJS. It is the default setting for yarn.

merrimanr replied:
What was the reason for rearranging the dependencies?  It's not a big deal but it's distracting and makes reviewing harder.

xyztdanid4 replied:
Yes. I was not enough carefully. Yarn did the rearrangement, but I can move the dependencies back to their original place if you would like to.

merrimanr replied:
No I was just curious.  If Yarn did this automatically we might as well leave it that way.  No need to revert.

merrimanr commented on metron-interface/metron-alerts/package.json:
```diff
@@ -27,8 +27,8 @@
     "moment": "^2.18.1",
-    "pikaday-time": "^1.6.1",
     "ng2-dragula": "^1.5.0",
+    "pikaday-time": "^1.6.1",
```
See comment for metron-config/package.json.

merrimanr commented:
I'm having trouble understanding what this PR actually does.  I see that package-lock.json was removed and yarn.lock was added.  Is that all that's needed to switch to yarn?  Do we need to also update our build commands to match those [here](https://yarnpkg.com/en/docs/usage)?

What considerations are there for development?  Are we supposed to be using yarn commands now instead of npm commands?  Do we need to install yarn locally?  We probably need to review our documentation and rpm build processes too.  We use npm all over the place.

xyztdanid4 commented:
@merrimanr

xyztdanid4 commented:
@merrimanr 
Yes this is the right way to switch to yarn.
Yes we have install yarn locally.
I did not know that we have so much dependency on npm. But for me it was hard to add new dependencies using npm, cause it was constantly failing. And yes I would recommend to use yarn, at least for the front end builds. What yarn does is that reduces the possibility of version mismatch  of the dependencies by importing the dependencies subdependency in the right version. So yes we have to use yarn commands if we merge this PR.

xyztdanid4 commented:
@merrimanr 
although it should work with npm as well, cause then it will use the package.json to install the dependencies.

merrimanr commented:
I think we are on board with switching to yarn so no concerns there.  There was a discussion on it but if anyone now feels differently, speak up.

Sounds like we still have a lot of work to do.  I would suggest reviewing the metron-config and metron-alerts READMEs and update them appropriately.  In fact, I would just search the whole project for npm commands and switch to yarn as needed.  You will definitely need to run this up in full dev and verify everything works.

Our maven front end plugin will also need to be reviewed and updated:  https://github.com/eirslett/frontend-maven-plugin.  We use that in metron-config and metron-alerts.

xyztdanid4 commented:
@merrimanr 
As we agreed yesterday I did the required changes in the pom xml-s.
I had to update the node dependency of the projects, otherwise the build was failed, cause one of the dependencies was dependent on node8.
There is no equivalent command to npm --prefix in yarn, so I updated the readme to just cd into that directory where the install step is required.
When we build the metron-alert project (angular4), there was an error, mainly because angular4 uses AOT as default builder strategy. I fixed this by addig env=prod to the package json.

sardell commented:
> What yarn does is that reduces the possibility of version mismatch of the dependencies by importing the dependencies subdependency in the right version

I believe npm also does this with the addition of [package-lock.json](https://docs.npmjs.com/files/package-lock.json).

nickwallen commented:
@sardell Thanks.

@xyztdanid4 Please let us know if you feel this PR is still relevant.  If not, please close the PR.  Otherwise, the PR may be closed as inactive according to the Metron Development Guidelines.

nickwallen commented:
-1 as this has become an abandoned PR.  I will open a request to close this.
